### PR TITLE
Add inline file delivery to WC_Download_Handler

### LINKED ISSUE #28410: Downloadable Products -- Inline file delivery
For our use case, we are using HTML files as virtual and downloadable products. We don't want our customers to have to download the html file and open it in the browser, we would rather have the file open directly in their browser when they try to download the content. We would like a setting in the Woocommerce -> Products -> Downloadable Products page that allows for inline file delivery. I plan on creating a PR for this next.

### PR DESCRIPTION
**Pull request by CaldwellYSR**
Create a new setting option for inline file delivery for downloadable
products. In the Download Handler create a new method for the new
setting as well as a method for returning the proper headers to deliver
files inline.

Resolves: #28410

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Creates a new setting for allowing downloadable product files to be delivered inline so they would open directly in the browser. This would be useful for stores selling HTML or PDF files so customers can access them directly without the need to download.

Only other change was aligning `=>` signs in an array to match [WordPress indentation coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#indentation).

Closes #28410 .

### How to test the changes in this Pull Request:

1. Create a downloadable/virtual product.
2. From Woocommerce Settings > Products > Downloadable Products select Inline File Delivery
3. Create a test purchase of the downloadable product
4. Test download delivery

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### CODE REVIEW

**Comment by CaldwellYSR** [comment]
Not sure what happened with the 4 commits. I only authored two commits. Any progress on this would be greatly appreciated. It's been working on our site for a while now but it's technical debt to keep having to apply this update every time we need to update the plugin.

**Review by vedanshujain — COMMENTED** [review]
Let's add this as a setting actually instead of a separate download method, since it can be used with X-Accel-Redirect as well, and not just with force downloads. 

We can then pass the value of this setting to `download_headers` and set the `Content-Disposition` header appropriately based on the params. That way, we won't have to create (and maintain) another download method, and another download header function. What do you think?

**Comment by CaldwellYSR** [comment]
> Let's add this as a setting actually instead of a separate download method, since it can be used with X-Accel-Redirect as well, and not just with force downloads.
> 
> We can then pass the value of this setting to `download_headers` and set the `Content-Disposition` header appropriately based on the params. That way, we won't have to create (and maintain) another download method, and another download header function. What do you think?

I agree with your suggestion and have made that change with the commit [386d75b](https://github.com/woocommerce/woocommerce/pull/28412/commits/386d75ba38f4a3399344c671320bdb2544ca34a8)

**Comment by CaldwellYSR** [comment]
@vedanshujain Any more suggestions for moving this forward?

**Comment by CaldwellYSR** [comment]
@vedanshujain I have been using this change for a while now and would love to get it merged into master so that I don't have to keep making the change every woocommerce update. Anything I can do or change to move this forward?

File: includes/admin/settings/class-wc-settings-products.php [hunk 1]
```diff
@@ -247,6 +247,16 @@ public function get_settings( $current_section = '' ) {
 						'autoload' => false,
 					),
 
+					array(
+						'title'         => __( 'Content-Disposition', 'woocommerce' ),
+						'desc'          => __( 'Deliver downloadable files inline', 'woocommerce' ),
+						'id'            => 'woocommerce_downloads_deliver_inline',
+						'type'          => 'checkbox',
+						'default'       => 'yes',
```
Comment (vedanshujain) [inline includes/admin/settings/class-wc-settings-products.php]: this should be disabled to false to keep the current behavior as it.

File: includes/class-wc-download-handler.php [hunk 2]
```diff
@@ -537,6 +543,14 @@ private static function clean_buffers() {
 		}
 	}
 
+	private static function get_content_disposition() {
```
Comment (vedanshujain) [inline includes/class-wc-download-handler.php:552]: this function is missing doc comments.

File: includes/abstracts/abstract-wc-order.php [hunk 3]
```diff
@@ -1633,6 +1633,7 @@ public function update_taxes() {
 				continue;
 			}
 			$saved_rate_ids[] = $tax->get_rate_id();
+			$tax->set_rate( $tax->get_rate_id() );
```
Comment (vedanshujain) [inline includes/abstracts/abstract-wc-order.php]: unintended change?

**Review by vedanshujain — COMMENTED** [review]
Looks like I accidentally unfollowed this thread and missed pings, sorry. I have left couple more comments, but I can take care of them myself as well, given that it has been a lot of time since the first round of review.

File: includes/abstracts/abstract-wc-order.php (hunk 3 above)
Comment (CaldwellYSR) [reply includes/abstracts/abstract-wc-order.php]: Definitely unintended. Not sure where it came from apologies.

**Comment by CaldwellYSR** [comment]
@vedanshujain I don't believe my fork is rebased to the upstream but I made the commented changes.

**Comment by vedanshujain** [comment]
> I don't believe my fork is rebased to the upstream but I made the commented changes.

No worries, I can take care of rebasing. Thanks for the patience, code looks fine, I am planning to do some more tests tomorrow before merging.

**Comment by CaldwellYSR** [comment]
> > I don't believe my fork is rebased to the upstream but I made the commented changes.
> 
> No worries, I can take care of rebasing. Thanks for the patience, code looks fine, I am planning to do some more tests tomorrow before merging.

Happy to hear it. Thanks again.

**Comment by vedanshujain** [comment]
Closing in favor of #31145 (rebased from trunk).

**Pull request closed without merging**